clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -6035,6 +6036,269 @@ test_kernel_switch_sequences()
+                }
+                if (memref.marker.type == TRACE_TYPE_MARKER &&
+                    memref.marker.marker_type == TRACE_MARKER_TYPE_SYSCALL_TRACE_START)
+                    in_syscall[i] = true;
+                if (in_syscall[i]) {
+                    // Test that syscall code is marked synthetic.
+                    assert(outputs[i]->is_record_synthetic());
+                    // Test that dynamically injected syscall code doesn't count toward
+                    // input ordinals, but does toward output ordinals.
+                    assert(outputs[i]->get_input_interface()->get_record_ordinal() ==
+                           prev_in_ord[i]);
+                    assert(outputs[i]->get_record_ordinal() > prev_out_ord[i]);
+                } else
+                    assert(!outputs[i]->is_record_synthetic());
+                if (type_is_instr(memref.instr.type))
+                    sched_as_string[i] += 'i';
+                else if (memref.marker.type == TRACE_TYPE_MARKER) {
+                    switch (memref.marker.marker_type) {
+                    case TRACE_MARKER_TYPE_VERSION: sched_as_string[i] += 'v'; break;
+                    case TRACE_MARKER_TYPE_TIMESTAMP: sched_as_string[i] += '0'; break;
+                    case TRACE_MARKER_TYPE_SYSCALL: sched_as_string[i] += 's'; break;
+                    case TRACE_MARKER_TYPE_SYSCALL_TRACE_END:
+                        in_syscall[i] = false;
+                        ANNOTATE_FALLTHROUGH;
+                    case TRACE_MARKER_TYPE_SYSCALL_TRACE_START:
+                        if (memref.marker.marker_value == SYSCALL_NUM_1)
+                            sched_as_string[i] += '1';
+                        else if (memref.marker.marker_value == SYSCALL_NUM_2)
+                            sched_as_string[i] += '2';
+                        else
+                            assert(false && "unknown system call num");
+                        break;
+                    default: sched_as_string[i] += '?'; break;
+                    }
+                }
+                prev_tid[i] = memref.instr.tid;
+                prev_in_ord[i] = outputs[i]->get_input_interface()->get_record_ordinal();
+                prev_out_ord[i] = outputs[i]->get_record_ordinal();
+            }
+        }
+        // Check the high-level strings.
+        for (int i = 0; i < NUM_OUTPUTS; i++) {
+            std::cerr << "cpu #" << i << " schedule: " << sched_as_string[i] << "\n";
+        }
+        // The instrs in the injected syscall sequence count towards the #instr
+        // quantum, but no context switch happens in the middle of the syscall seq.
+        assert(sched_as_string[0] ==
+               "Av0is1ii1,Cv0is1ii1,Aiis2iii2,Ciis2iii2,Aiis1ii1,Ciis1ii1,Aiis2iii2,"
+               "Ciis2iii2,Aiis1ii1,Ciis1ii1");
+        assert(sched_as_string[1] ==
+               "Bv0is1ii1iis2iii2iis1ii1iis2iii2iis1ii1____________________________"
```

Comment:
@derekbruening: Should core 2 have stolen either A or C from core 1?

Reply:
There is a migration threshold that is almost certainly what is preventing that: it's assumed the cost of A migrating and having cold caches is higher than waiting to run on core0.

Reply:
Ack.